src/main/java/betterquesting/questing/tasks/TaskFluid.java

```diff
@@ -81,115 +85,217 @@ public void onInventoryChange(@Nonnull DBEntry<IQuest> quest, @Nonnull Participa
+            indexedContainer.updateFluidContainer(handler, itemsToConsume, consume);
+            totalDrained += amountFullDrained;
+            remaining -= amountFullDrained;
+            if (remaining <= 0) {
+                break;
             }
         }
+        return totalDrained;
+    }
+
+    @Nonnull
+    private int[] updateBulkProgress(int[] playerProgress, ParticipantInfo pInfo) {
+        List<UUID> uuidsToUpdate = consume ? Collections.singletonList(pInfo.UUID) : pInfo.ALL_UUIDS;
 
-        if (updated) setBulkProgress(progress);
-        checkAndComplete(pInfo, quest, updated);
+        int[] updatedProgress = updateUserProgress(pInfo.UUID, playerProgress);
+        for (UUID uuid : uuidsToUpdate) {
+            if (uuid == pInfo.UUID) continue;
+            updateUserProgress(uuid, playerProgress);
+        }
+        return updatedProgress;
     }
 
-    private void checkAndComplete(ParticipantInfo pInfo, DBEntry<IQuest> quest, boolean resync) {
-        final List<Tuple<UUID, int[]>> progress = getBulkProgress(consume ? Collections.singletonList(pInfo.UUID) : pInfo.ALL_UUIDS);
-        boolean updated = resync;
+    /**
+     * Update the task progress for given user
+     *
+     * @param userUUID      the user's id
+     * @param progressIn    the progress to merge, treated as immutable
+     * @return the updated task progress
+     */
+    private int[] updateUserProgress(UUID userUUID, int[] progressIn) {
+        return userProgress.merge(userUUID, progressIn, (existingProgress, progressToMerge) -> {
+            Preconditions.checkArgument(existingProgress.length == progressToMerge.length
+                            && progressToMerge.length == requiredFluids.size(),
+                    "Lengths of user's known progress and new detected progress to merge don't match!");
+            if (groupDetect) {
+                progressChanged = true;
+                return progressToMerge;
+            }
 
-        topLoop:
-        for (Tuple<UUID, int[]> value : progress) {
-            for (int j = 0; j < requiredFluids.size(); j++) {
-                if (value.getSecond()[j] >= requiredFluids.get(j).amount) continue;
-                continue topLoop;
+            for (int i = 0; i < existingProgress.length; i++) {
+                // Skip if already fulfilled
+                if (existingProgress[i] >= requiredFluids.get(i).amount) continue;
+
+                if (consume) {
+                    // Make sure we keep the progress that has already consumed stuff before
+                    existingProgress[i] += progressToMerge[i];
+                    progressChanged = true;
+                }
+                else {
+                    if (existingProgress[i] != progressToMerge[i]) {
+                        progressChanged = true;
+                    }
+                    // Otherwise the progressIn overwrites the current progress
+                    existingProgress[i] = progressToMerge[i];
+                }
```

Comment:
i believe this could be
```suggestion
                else if (existingProgress[i] != progressToMerge[i]) {
                    progressChanged = true;
                    // Otherwise the progressIn overwrites the current progress
                    existingProgress[i] = progressToMerge[i];
                }
```
since they are either not equal, and then the flag is enabled and they are made equal, or they are equal and set to themself.